### DISCUSSION TITLE
more robust joystic/HOTAS support

### DISCUSSION THREAD

**Discussion by m9365428** [sezanzeb/input-remapper#606] [Ideas]
TLDR: remap analog axis on joystick to virtual device for game comparability. (aka replicate [UCR](https://github.com/Snoothy/UCR) type features in linux)

I just purchased a Virpil HOTAS and have discovered an apparently well known issue in the "SIM" world of no 2 games implementing inputs or comparability the same way. 
I rapidly found [UCR](https://github.com/Snoothy/UCR) which allows you to remap your inputs to a virtual joystick. AKA virtual joystick matches what devs want and UCR converts inputs from your joystick to match. Only one problem. Its windows only!

So after looking for a month now I find your program and its wonderful. And the closest thing I have found yet to what I need i have yet to find. So here is the ask, can the following feature even be considered?
1. add virtual output devices
2. remap analog input axis

Detail:
I used to play (getting back into) flight, combat, and space sims back in the day so that is where all the examples are coming from.
1. As an example let say accelerate is 'w' deploy flare is 'ctrl + c' and spool hyper drive is 'ctrl + v + w' (some games have way too many combos. 
If someone used this software to remap 'numpad 7' to 'ctrl + c' and while holding 'w' pressed 'numpad 7'. The system would spool hyperdrive. If however 'numpad 7' send the command from a virtual keyboard with a different hardware id then you would deply flare while accelerating.
Last time I played these kind of games was on windows and it sucked because of how they handle HID's. AKA all keyboards are threated as the same. And the only way to get the above mentioned functionaliy was to install Auto Hotkey and a Input Driver that i dont remember the name of at the moment. And we get this kind of functionality on Linux basically for free because of HID handling.
(sorry ranted there) but allowing a virtual keyboard would in effect add the ability to issue 2 separate input streams to games without them interfering with each other.

2. Long and short is that my joysticks use a strange axis naming scheme and as a result most games don't recognize one or more axis as the driver name rX isn't recognized or 5 analogs on 1 stick is too many. Most of these games are compatible with combined inputs from multiple devices at once but not if they have more than 3 axis.
In short being able to rename/remap the analog axis would allow compatibility fixes.

I know this is alot but honestly this is the only linux related project I have seen that even comes close to what I need. I have no right to ask these of you but my only other choice is gonna be a software/programmer bounty.

**Comment by jonasBoss** [comment]
>  2. remap analog input axis

I implemented that possibility on the `beta` branch, which should be released as version 2.0 at some point in the next months. You can use the beta branch if you install it manually.

>  1. add virtual output devices

That is not yet supported, but "easy" to implement. Virtual devices are hard coded here: https://github.com/sezanzeb/input-remapper/blob/9adc300aeefe39c94655738ed52e848808f991f3/inputremapper/injection/global_uinputs.py#L31
You can just add a new device or change a existing device to fit your needs.
If you add a new device you also want to add the name to the `KnownUinput` enum: https://github.com/sezanzeb/input-remapper/blob/9adc300aeefe39c94655738ed52e848808f991f3/inputremapper/configs/mapping.py#L78
otherwise mappings will not work.

**Reply by m9365428** [reply]
Thank you for the prompt reply. 
Awesome, I will start testing on beta branch and look forward to full release.

I will dig in and see if I can figure out the virtual outputs. Thanks for the starting points.
